Change PathInternal.IsCaseSensitive to a constant

lambdageek opened the pull request:
See https://github.com/dotnet/runtime/issues/54313#issuecomment-863347311 and https://github.com/dotnet/maui/issues/822#issuecomment-863323985

The code that initialized `System.IO.PathInternal.s_isCaseSensitive` did I/O in the temp folder to check for case sensitivity.  This is a startup perf hit on Android where low-end devices may have very slow storage. It also returns results that may be misleading since different paths on Unix may have different filesystems mounted that have different case sensitivity (a common case on Android is ext4 for internal storage and exFAT for the SD card).

Instead, return a constant value based on the common platform behavior - case insensitive on Windows and Apple platforms, case sensitive elsewhere.

Also remove `Path.IsCaseSensitive` and `Path.StringComparison` - callers should use `PathInternal` properties directly.

Fixes https://github.com/dotnet/runtime/issues/54339

jkotas commented:
Would this make sense to hardcode this for all mobile targets ? (e.g. have two files - `*.CaseSensitive.cs` and `*.CaseInsensitive.cs` and include them as appropriate).

marek-safar commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.Android.cs:
```diff
@@ -0,0 +1,23 @@
+        /// filesystems may or may not be case sensitive.  On Android, internal storage is usually
+        /// case-sensitive ext4, while SD card storage is case-insensitive exFAT.  As a rough
+        /// guideline we will assume case sensitivity.
+        internal static bool IsCaseSensitive { get => true; }
```
Simpler syntax is `internal static bool IsCaseSensitive => true;`

marek-safar reviewed (commented):
I think it would make sense to hardcode this for known behaviour (e.g. iOS like or browser-wasm).

I'm not sure why there is similar logic in Path class as well https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L961

jkotas commented:
> I'm not sure why there is similar logic in Path class as well

The behavior is actually hardcoded in Path class for all platforms:

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/IO/Path.Unix.cs#L137

The reason for this duplication is that there is no public `Path.IsCaseSensitive` property, but this property is needed by outside of CoreLib. PathInternal is included by outside of CoreLib to get access to `IsCaseSensitive` and several other internal properties.

We should make the logic in PathInternal to be hardcoded in the same way as the logic in Path. It helps with both performance and consistency. The check in PathInternal will need to be dynamic (using `OperatingSystem.Is*`).

lambdageek commented:
I'm following along with most of the discussion, but I don't understand this part @jkotas:

>  The check in PathInternal will need to be dynamic (using OperatingSystem.Is*).

Aren't we talking about doing the same thing in PathInternal as in Path: just hardcode based on the common behavior of the underlying OS?  Why would it need to be a dynamic check if we know which RID we're targeting?

marek-safar commented:
@lambdageek because the file is shared as a compilation file you cannot assume that the assembly where the file is included is multi-targeting (has RID specific builds). OperatingSystem properties are static constants so I don't think there will be any perf hit.

lambdageek committed:
Return constants in PathInternal.GetIsCaseSensitive() on mobile platforms

In particular on Android probing using I/O is slow and contributes to slow app startup.

Fixes https://github.com/dotnet/runtime/issues/54339

lambdageek commented:
Ok, updated to return constants for some known `OperatingSystem.IsXYZ` values.  I left the probing in place as the last case (so desktop Linux, Windows, FreeBSD and "none of the above" will still do the I/O probing)

lambdageek committed:
Put OperatingSystem probing under a NET6_0_OR_GREATER define

jkotas commented:
I do not think we should be doing any probing. The IsCaseInsensitive property on Path is not doing any probing, and it better be in sync with the property on PathInternal. The probing on PathInternal can only ever lead to inconsistent behavior.

jkotas commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -31,13 +31,30 @@ internal static StringComparison StringComparison
         /// </summary>
+        private static bool GetIsCaseSensitive()
+        {
+#if NET6_0_OR_GREATER
+            // Return a constant for mobile platforms without resorting to I/O
+            if (OperatingSystem.IsMacOS() || OperatingSystem.IsMacCatalyst() || OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsWatchOS())
+                return true;
+            if (OperatingSystem.IsBrowser())
+                return false;
+            if (OperatingSystem.IsAndroid())
+                return false;
+#endif
+            return GetIsCaseSensitiveByProbing();
```
```suggestion
#if MS_IO_REDIST
            return false; // Windows is always case-insensitive
#else
    return OperatingSystem.IsWindows() || OperatingSystem.IsMacOS() || OperatingSystem.IsMacCatalyst() || OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsWatchOS();
#endif
```

marek-safar replied:
If no probing is needed could we remove the implementation from Path and call into PathInternal all the time?

jkotas replied:
Yes, that is a good idea. We can even delete the property on Path and change the few places that use it to call PathInternal instead.

jkotas commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -31,13 +31,30 @@ internal static StringComparison StringComparison
+#if NET6_0_OR_GREATER
+            // Return a constant for mobile platforms without resorting to I/O
+            if (OperatingSystem.IsMacOS() || OperatingSystem.IsMacCatalyst() || OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsWatchOS())
+                return true;
```
I think this is backwards - it should return false for MacOS.

jkotas commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -31,13 +31,30 @@ internal static StringComparison StringComparison
+        {
+#if NET6_0_OR_GREATER
+            // Return a constant for mobile platforms without resorting to I/O
+            if (OperatingSystem.IsMacOS() || OperatingSystem.IsMacCatalyst() || OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsWatchOS())
```
The ifdef in Path.Unix.cs does not have WatchOS. Is it a bug?

adamsitnik commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -31,13 +31,30 @@ internal static StringComparison StringComparison
         /// cases where, for example, different file systems are mounted with differing sensitivities.
         /// </remarks>
-        private static bool GetIsCaseSensitive()
+        private static bool GetIsCaseSensitiveByProbing()
```
I think that it would be great to use the old implementation as a test of the new one.

You could introduce a new test in [System.IO.FileSystem.Tests.csproj](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj) include the `.cs` file like this:


```xml
    <Compile Include="$(CommonPath)System\IO\PathInternal.CaseSensitivity.cs"
             Link="Common\System\IO\PathInternal.CaseSensitivity.cs" />
```

 and move `GetIsCaseSensitiveByProbing` there to get the "expected".

Something like this:

```cs
[Fact]
public void GetIsCaseSensitiveReturnsSameValueAsProbing()
{
    Assert.Equal(GetIsCaseSensitiveByProbing(), PathInternal.IsCaseSensitive);
}
```

jkotas replied:
That would be unreliable test. The temp directory on Unix may live on a case insensensitive volume.

adamsitnik replied:
> That would be unreliable test. The temp directory on Unix may live on a case insensensitive volume.

So perhaps it should be using current directory instead of temp? I know that it would still be far from perfect but I would prefer to have a single non-perfect test rather than no tests at all.

jkotas replied:
Current directory has the same problem. We can have a test like that, but it should be marked as outerloop.

adamsitnik replied:
> We can have a test like that, but it should be marked as outerloop.

Sounds good to me.

lambdageek replied:
> So perhaps it should be using current directory instead of temp

That might fail on mobile for trivial reasons - I'm not sure the tests run from a writable directory. I'll give it a try and see.

lambdageek committed:
Apply suggested changes

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

lambdageek committed:
Implement Path.IsCaseSensitive as PathInternal.IsCaseSensitive

Also Path.StringComparison => PathInternal.StringComparison

lambdageek committed:
Add test for PathInternal.IsCaseSensitive

Move GetIsCaseSensitiveByProbing to FileSystemTest

jkotas commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -31,29 +31,13 @@ internal static StringComparison StringComparison
-        /// and then tests for its existence with lower-case letters.  This could return invalid results in corner
-        /// cases where, for example, different file systems are mounted with differing sensitivities.
-        /// </remarks>
         private static bool GetIsCaseSensitive()
```
This should be `internal static bool IsCaseSensitive` property and `s_isCaseSensitive` should be deleted. It is not profitable to cache it in a static, and it prevents the linker constant propagation from kicking in.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:
```diff
@@ -957,11 +957,11 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
         }
 
+        /// <summary>Gets whether the system is case-sensitive.</summary>
+        internal static bool IsCaseSensitive => PathInternal.IsCaseSensitive;
```
I would change the few places that use this to call `PathInternal.IsCaseSensitive` directly instead of introducing this wrapper.

jkotas commented on src/libraries/System.IO.FileSystem/tests/PathInternalTests.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class PathInternalTests : FileSystemTest
+    {
+        [Fact]
```
```suggestion
        [Fact]
        [OuterLoop]
```

jkotas commented on src/libraries/System.IO.FileSystem/tests/FileSystemTest.cs:
```diff
@@ -127,5 +127,33 @@ protected void ReadOnly_FileSystemHelper(Action<string> testAction, string subDi
+            }
+            catch
+            {
+                // In case something goes wrong (e.g. temp pointing to a privilieged directory), we don't
```
This `catch` made sense when it was part of the shipping code. Now that it is test, I think it would be better to delete it. We want to know when this fails as a test.

lambdageek committed:
Drop PathInternal.s_isCaseSensitive cache field

lambdageek committed:
Delete Path.IsCaseSensitive and Path.StringComparison

update callers to use PathInternal.IsCaseSensitive and PathInternal.StringComparison

lambdageek committed:
Remove catch clause from GetIsCaseSensitiveByProbing

tannergooding commented on src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs:
```diff
@@ -12,47 +12,27 @@ namespace System.IO
-                // want to fail just because of a casing test, so we assume case-insensitive-but-preserving.
-                return false;
+#if MS_IO_REDIST
+                return false; // Windows is always case-insensitive
```
Is this going to break if I pass `FILE_FLAG_POSIX_SEMANTICS` to `CreateFile` or if I modify a folder to set it as case insensitive using `fsutil` (which is supported since Win10 17134)?

There are multiple ways to force windows to be case sensitive, including for the WSL mounted drive, which is fully browsable via File explorer and can be interacted with via the file system APIs

tannergooding replied:
https://devblogs.microsoft.com/commandline/per-directory-case-sensitivity-and-wsl/ goes into some more details about various changes that went in, but even prior to this NTFS has always supported case sensitivity. Its just that Windows hadn't handled it as "first class" if you enabled it.

lambdageek replied:
It will be just as broken as before (when we probed the Temp folder, which is not a good proxy for the behavior of a _specific_ directory - neither on Unix nor on Windows), but faster.

jkotas replied:
This has been a global setting, independent on the path. This logic is broken if you have parts of the file system configured differently. This PR does not change it. Yes, it would be nice to fix that.

The unmanaged side of the runtime has similar global setting. It is larger problem than just what you can see in the managed libraries.

lambdageek committed:
fixup Net5CompatTests

lambdageek committed:
Mark new test [OuterLoop]

adamsitnik commented on src/libraries/System.IO.FileSystem/tests/FileSystemTest.cs:
```diff
@@ -127,5 +127,24 @@ protected void ReadOnly_FileSystemHelper(Action<string> testAction, string subDi
+        /// and then tests for its existence with lower-case letters.  This could return invalid results in corner
+        /// cases where, for example, different file systems are mounted with differing sensitivities.
+        /// </remarks>
+        protected static bool GetIsCaseSensitiveByProbing(string probingDirectory)
+        {
+	    string pathWithUpperCase = Path.Combine(probingDirectory, "CASESENSITIVETEST" + Guid.NewGuid().ToString("N"));
+	    using (new FileStream(pathWithUpperCase, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 0x1000, FileOptions.DeleteOnClose))
+	    {
+		string lowerCased = pathWithUpperCase.ToLowerInvariant();
+		return !File.Exists(lowerCased);
+	    }
+        }
```
formatting

```suggestion
    protected static bool GetIsCaseSensitiveByProbing(string probingDirectory)
    {
        string pathWithUpperCase = Path.Combine(probingDirectory, "CASESENSITIVETEST" + Guid.NewGuid().ToString("N"));
        using (new FileStream(pathWithUpperCase, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 0x1000, FileOptions.DeleteOnClose))
        {
            string lowerCased = pathWithUpperCase.ToLowerInvariant();
            return !File.Exists(lowerCased);
        }
    }
```

adamsitnik reviewed (approved):
LGTM, thank you @lambdageek !

adamsitnik committed:
Apply suggestions from code review

adamsitnik committed:
fix formatting

lambdageek commented:
Should we backport this to .NET 6 Preview 6?  (It's not critical for MAUI on Android - they mitigated the perf hit on their end)

stephentoub commented on src/libraries/System.IO.FileSystem/tests/PathInternalTests.cs:
```diff
@@ -0,0 +1,18 @@
+        [OuterLoop]
+        public void PathInternalIsCaseSensitiveMatchesProbing()
+        {
+            var probingDirectory = TestDirectory;
```
```suggestion
            string probingDirectory = TestDirectory;
```

lambdageek committed:
Apply suggestions

Co-authored-by: Stephen Toub <stoub@microsoft.com>

stephentoub reviewed (approved):
Thanks.

stephentoub commented:
> Should we backport this to .NET 6 Preview 6?

If the perf issue was mitigated, I don't think we need to.

lambdageek commented:
/azp run runtime-staging runtime-dev-innerloop dotnet-linker-tests
